src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -208,7 +210,40 @@ public TextWriter EndWritingScope()
                 _originalWriter = null;
             }
 
-            return writer;
+            var tagHelperContentWrapperTextWriter = new TagHelperContentWrapperTextWriter(Encoding.UTF8);
+            var razorWriter = writer as RazorTextWriter;
+            if (razorWriter != null)
+            {
+                razorWriter.CopyTo(tagHelperContentWrapperTextWriter);
+            }
+            else
+            {
+                tagHelperContentWrapperTextWriter.Write(writer.ToString());
+            }
+
+            return tagHelperContentWrapperTextWriter;
+        }
+
+        /// <summary>
+        /// Writes a type that implements <see cref="ITextWriterCopyable"/> to the respective <see cref="Output"/>.
+        /// </summary>
+        /// <param name="copyableTextWriter">Contains the data to be written.</param>
+        public void Write(ITextWriterCopyable copyableTextWriter)
+        {
+            WriteTo(Output, copyableTextWriter);
+        }
+
+        /// <summary>
+        /// Writes a type that implements <see cref="ITextWriterCopyable"/> to the specified <see cref="TextWriter"/>.
+        /// </summary>
+        /// <param name="writer">The <see cref="TextWriter"/> to which the output must be written to.</param>
+        /// <param name="copyableTextWriter">Contains the data which needs to be written to the output.</param>
+        public void WriteTo(TextWriter writer, ITextWriterCopyable copyableTextWriter)
```

Comment:
as mentioned in aspnet/Razor#312, these methods must be named `WriteLiteralTo` because they perform no HTML encoding.